### ISSUE TITLE
respond method can't use additional model in gson

### ISSUE THREAD

**Issue by nobeans** [grails/grails-core#10997]
### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Prepare actions like this.

    ```
    // in BookController
    class BookController {
        //...
        def show(Long id) {
            // There are both show.gson and show.gsp.
            respond Book.get(id), model: [foo: "FOO"]
        }
        def show2(Long id) {
            // There is only show3.gson.
            render view: 'show2', model: [book: Book.get(id), foo: "FOO"]
        }
        def show3(Long id) {
            // There are both show2.gson and show2.gsp.
            render view: 'show3', model: [book: Book.get(id), foo: "FOO"]
        }
        //...
    }
    ```

2. Prepare GSON and GSP like this.

    ```
    // in show.gson
    model {
        Book book
        String foo
    }
    json {
        book g.render(book)
        foo foo
    }
    ```

    Copy show.gson to show2.gson and show3.gson.
    Copy show.gsp to show3.gsp.

2. Run application.

3. Access to the actions.

### Expected Behaviour

```
> curl http://localhost:8888/book/show/1.json
{"book":{"id":1,"title":"Sample Book","author":"John"},"foo":"FOO"}
```

```
> curl http://localhost:8888/book/show2/1.json
{"book":{"id":1,"title":"Sample Book","author":"John"},"foo":"FOO"}
```

```
> curl http://localhost:8888/book/show3/1.json
{"book":{"id":1,"title":"Sample Book","author":"John"},"foo":"FOO"}
```

### Actual Behaviour

```
> curl http://localhost:8888/book/show/1.json
{"book":{"id":1,"title":"Sample Book","author":"John"},"foo":null}
--> (1) The 'model' map's entry couldn't be used in gson.
```

```
> curl http://localhost:8888/book/show2/1.json
{"book":{"id":1,"title":"Sample Book","author":"John"},"foo":"FOO"}
--> Expectedly.
```

```
> curl http://localhost:8888/book/show3/1.json
<!doctype html>
<html lang="en" class="no-js">
<head>
......
-> (2) Even when I specified 'json' format, the response was HTML.
```

(1) and (2) may be different issues.

### Environment Information

- **Operating System**: macOS 10.13.4
- **Grails Version:** 3.3.5
- **JDK Version:** 1.8.0_161

### Example Application

- https://github.com/nobeans/grails-issues/tree/master/g335-gson-additional-model

**Comment by nobeans** [comment]
> model - The model to use in case of HTML rendering

Oh. It was been describing surely. So that, my expected behavior is "Improvement" request, isn't it?

**Comment by nobeans** [comment]
OK, I understood that (1) is the correct behavior. How about (2)?

**Comment by puneetbehl** [comment]
@nobeans You can pass additional model as:
```
 Map model = [book:Book.get(id), foo: "FOO"]
 respond(model)
```
or
```
 Map args = [view: 'show']
 Map model = [book:Book.get(id), foo: "FOO"]
 respond(args, model)
```

Also, for (2) you are using `render` method which I think does not take care of Content Negotiation. You could use something like `withFormat` as described [here](https://docs.grails.org/latest/ref/Controllers/withFormat.html)

**Comment by puneetbehl** [comment]
FYI, I am still working on making `respond Book.get(id), model: [foo: "FOO"]` to work for JSON views but to me following is more clear:
```
Map model = [book:Book.get(id), foo: "FOO"]
respond(model)
```

**Comment by nobeans** [comment]
@puneetbehl

Wow, the two ways certainly work even in Grals 3.3.5. Great! And I'm very glad to see your PR which supports `respond(domain, model)` notation.


> Also, for (2) you are using render method which I think does not take care of Content Negotiation.

When there are both `show3.gson` and `show3.gsp` in the above sample, `render` can use `show3.gson` for the request to `http://localhost:8080/book/show2/1.json`. So I thought `render` takes care of Content Negotiation in a way... It's a very limited support, isn't it?


> You could use something like withFormat as described here

Yes, actually I've workarounded this problem by `withFormat` like this:

```groovy
Map model = [book:Book.get(id), foo: "FOO"]
withFormat {
    '*' { render view: 'show5', model: model } // default=JSON
    html { render view: 'show5_', model: model }
}
```

It works expectedly when there are `show5.gson` and `show5_.gsp`.

However, if the two base names are same like `show5.gson` and `show5.gsp`, it doesn't work. The response is HTML format as same as the result of `show3`. It seems another issue in there.

Updated my sample project: https://github.com/nobeans/grails-issues/tree/master/g335-gson-additional-model
